File: backend/test/safebrowser/ParentUnblockRequestService.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import { BlockDecisionStateService, InMemoryBlockDecisionStateRepository } from '../../dist/safebrowser/BlockDecisionStateStore.js';
import {
  InMemoryParentUnblockRequestRepository,
  ParentUnblockRequestService,
  UnblockRequestError,
} from '../../dist/safebrowser/ParentUnblockRequestService.js';

function decision(overrides = {}) {
  return {
    domain: 'blocked.example',
    outcome: 'BLOCK',
    source: 'PARENT_DENYLIST',
    reasonCode: "blocked by your family's block list",
    coverage: 'DOMAIN_ONLY',
    ...overrides,
  };
}

async function setup(decisionOverrides = {}) {
  const decisions = new InMemoryBlockDecisionStateRepository();
  const decisionService = new BlockDecisionStateService(decisions, () => new Date('2026-01-01T00:00:00Z'));
  const built = decision(decisionOverrides);
  const state = await decisionService.record('fam-1', 'prof-1', `https://${built.domain}/`, null, built);
  const requests = new InMemoryParentUnblockRequestRepository();
  const service = new ParentUnblockRequestService(requests, decisions, () => new Date('2026-01-01T00:00:00Z'));
  return { state, requests, service };
}

test('submit creates a PENDING request tied to the block decision', async () => {
  const { state, service } = await setup();
  const request = await service.submit('fam-1', 'prof-1', state.id);
  assert.equal(request.status, 'PENDING');
  assert.equal(request.blockDecisionId, state.id);
  assert.equal(request.domain, 'blocked.example');
});

test('submit rejects a request for a non-requestable (security) block', async () => {
  const { state, service } = await setup({ domain: 'malware.example', source: 'SECURITY_DENYLIST' });
  await assert.rejects(
    () => service.submit('fam-1', 'prof-1', state.id),
    (err) => err instanceof UnblockRequestError && err.code === 'NOT_REQUESTABLE',
  );
});

test('submit rejects a second request while one is already pending', async () => {
  const { state, service } = await setup();
  await service.submit('fam-1', 'prof-1', state.id);
  await assert.rejects(
    () => service.submit('fam-1', 'prof-1', state.id),
    (err) => err instanceof UnblockRequestError && err.code === 'ALREADY_PENDING',
  );
});

test('approvePermanent moves PENDING to APPROVED_PERMANENT', async () => {
  const { state, service } = await setup();
  const request = await service.submit('fam-1', 'prof-1', state.id);
  const decided = await service.approvePermanent(request.id);
  assert.equal(decided.status, 'APPROVED_PERMANENT');
  assert.notEqual(decided.decidedAt, null);
});

test('approveTemporary sets an expiry and rejects a non-positive duration', async () => {
  const { state, service } = await setup();
  const request = await service.submit('fam-1', 'prof-1', state.id);
  const decided = await service.approveTemporary(request.id, 3600_000);
  assert.equal(decided.status, 'APPROVED_TEMPORARY');
  assert.equal(decided.temporaryApprovalExpiresAt.getTime(), new Date('2026-01-01T01:00:00Z').getTime());

  const { state: state2, service: service2 } = await setup();
  const request2 = await service2.submit('fam-1', 'prof-1', state2.id);
  await assert.rejects(
    () => service2.approveTemporary(request2.id, -1),
    (err) => err instanceof UnblockRequestError && err.code === 'INVALID_DURATION',
  );
});

test('a decided request cannot be decided again', async () => {
  const { state, service } = await setup();
  const request = await service.submit('fam-1', 'prof-1', state.id);
  await service.deny(request.id);
  await assert.rejects(
    () => service.approvePermanent(request.id),
    (err) => err instanceof UnblockRequestError && err.code === 'ILLEGAL_TRANSITION',
  );
});

test('submit rejects a reference to a nonexistent block decision', async () => {
  const { service } = await setup();
  await assert.rejects(
    () => service.submit('fam-1', 'prof-1', 'does-not-exist'),
    (err) => err instanceof UnblockRequestError && err.code === 'DECISION_NOT_FOUND',
  );
});

File: android/app/src/test/java/org/pca/app/feature/webprotection/safebrowser/ParentUnblockRequestServiceTest.kt
package org.pca.app.feature.webprotection.safebrowser

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.pca.app.feature.webprotection.policy.WebDecisionOutcome
import org.pca.app.feature.webprotection.policy.WebDecisionSource
import org.pca.app.feature.webprotection.policy.WebReasonId
import org.pca.app.foundation.InMemoryPersistentStateStore

private fun heldDecision(requestable: Boolean = true, id: String = "decision-1"): BlockDecisionState = BlockDecisionState(
    id = id,
    familyId = "family-1",
    profileId = "profile-1",
    domain = "blocked.example",
    url = "https://blocked.example/page",
    pageTitle = "Blocked",
    outcome = WebDecisionOutcome.BLOCK,
    source = WebDecisionSource.PARENT_DENYLIST,
    reasonId = WebReasonId.PARENT_DENYLIST,
    reasonCode = "Blocked by a parent rule",
    requestable = requestable,
    createdAtEpochMillis = 0L,
)

class ParentUnblockRequestServiceTest {

    private fun service() = ParentUnblockRequestService(PersistentParentUnblockRequestRepository(InMemoryPersistentStateStore()))

    @Test
    fun `submitting a requestable decision creates a PENDING request, never self-approved`() {
        val request = service().submit(heldDecision())

        assertEquals(ParentUnblockRequestStatus.PENDING, request.status)
        assertEquals("blocked.example", request.domain)
    }

    @Test
    fun `a non-overridable security decision cannot be submitted at all -- doc 27's non-requestable rule`() {
        try {
            service().submit(heldDecision(requestable = false))
            assertTrue("expected DecisionNotRequestable", false)
        } catch (_: UnblockRequestError.DecisionNotRequestable) {
            // expected
        }
    }

    @Test
    fun `a duplicate submission for the same decision is rejected as already pending`() {
        val svc = service()
        svc.submit(heldDecision())

        try {
            svc.submit(heldDecision())
            assertTrue("expected AlreadyPending", false)
        } catch (_: UnblockRequestError.AlreadyPending) {
            // expected
        }
    }

    @Test
    fun `approveTemporary transitions to APPROVED_TEMPORARY with a real expiry, never a self-approval path`() {
        val svc = service()
        val request = svc.submit(heldDecision())

        val approved = svc.approveTemporary(request.id, durationMs = 60_000L)

        assertEquals(ParentUnblockRequestStatus.APPROVED_TEMPORARY, approved.status)
        assertTrue(approved.temporaryApprovalExpiresAtEpochMillis!! > 0L)
    }

    @Test
    fun `a decided request cannot be decided again -- illegal transition`() {
        val svc = service()
        val request = svc.submit(heldDecision())
        svc.deny(request.id)

        try {
            svc.approvePermanent(request.id)
            assertTrue("expected IllegalTransition", false)
        } catch (_: UnblockRequestError.IllegalTransition) {
            // expected
        }
    }

    @Test
    fun `requests survive a fresh service instance over the same backing store`() {
        val backing = InMemoryPersistentStateStore()
        val first = ParentUnblockRequestService(PersistentParentUnblockRequestRepository(backing))
        val request = first.submit(heldDecision())

        val secondRepo = PersistentParentUnblockRequestRepository(backing)
        assertEquals(ParentUnblockRequestStatus.PENDING, secondRepo.get(request.id)?.status)
    }
}
